File: backend/Kurator.Core/Entities/Watchlist.cs
using Kurator.Core.Enums;

namespace Kurator.Core.Entities;

/// <summary>
/// Реестр потенциальных угроз (Watchlist)
/// ВАЖНО: Данные НЕ шифруются (доступ только у администраторов и аналитиков угроз)
/// </summary>
public class Watchlist
{
    public int Id { get; set; }

    /// <summary>
    /// ФИО / псевдоним (НЕ шифруется)
    /// </summary>
    public string FullName { get; set; } = string.Empty;

    /// <summary>
    /// Роль / статус (например: журналист, бывший партнёр)
    /// </summary>
    public string? RoleStatus { get; set; }

    /// <summary>
    /// ID сферы риска из справочника
    /// </summary>
    public int? RiskSphereId { get; set; }

    /// <summary>
    /// Источник угрозы
    /// </summary>
    public string? ThreatSource { get; set; }

    /// <summary>
    /// Дата возникновения конфликта
    /// </summary>
    public DateTime? ConflictDate { get; set; }

    /// <summary>
    /// Уровень риска
    /// </summary>
    public RiskLevel RiskLevel { get; set; }

    /// <summary>
    /// Периодичность мониторинга
    /// </summary>
    public MonitoringFrequency MonitoringFrequency { get; set; }

    /// <summary>
    /// Дата последней проверки
    /// </summary>
    public DateTime? LastCheckDate { get; set; }

    /// <summary>
    /// Дата следующей проверки
    /// </summary>
    public DateTime? NextCheckDate { get; set; }

    /// <summary>
    /// Ход / динамика (текущая ситуация)
    /// </summary>
    public string? DynamicsDescription { get; set; }

    /// <summary>
    /// Ответственный наблюдатель (аналитик угроз)
    /// </summary>
    public int? WatchOwnerId { get; set; }

    /// <summary>
    /// JSON массив с путями к прикрепленным материалам (НЕ шифруются)
    /// </summary>
    public string? AttachmentsJson { get; set; }

    /// <summary>
    /// Статус записи (активный / деактивированный)
    /// </summary>
    public bool IsActive { get; set; } = true;

    /// <summary>
    /// Дата создания
    /// </summary>
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    /// <summary>
    /// Дата последнего обновления
    /// </summary>
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;

    /// <summary>
    /// ID пользователя, внёсшего последнее изменение
    /// </summary>
    public int UpdatedBy { get; set; }

    // Navigation properties
    public virtual User? WatchOwner { get; set; }
    public virtual User UpdatedByUser { get; set; } = null!;
    public virtual ReferenceValue? RiskSphere { get; set; }
    public virtual ICollection<WatchlistHistory> History { get; set; } = new List<WatchlistHistory>();
}

File: frontend/lib/schemas/watchlist.ts
/**
 * Watchlist Zod schemas
 */
import { z } from 'zod';
import { createPaginatedResponseSchema, nullableDateStringSchema } from './common';

// Risk level enum
export const riskLevelSchema = z.enum(['Low', 'Medium', 'High', 'Critical']);
export type RiskLevel = z.infer<typeof riskLevelSchema>;

// Monitoring frequency enum
export const monitoringFrequencySchema = z.enum(['Weekly', 'Monthly', 'Quarterly', 'AdHoc']);
export type MonitoringFrequency = z.infer<typeof monitoringFrequencySchema>;

// Watchlist entry schema
export const watchlistEntrySchema = z.object({
  id: z.number().int().positive(),
  fullName: z.string(),
  roleStatus: z.string().nullable().optional(),
  riskSphereId: z.number().nullable().optional(),
  threatSource: z.string().nullable().optional(),
  conflictDate: nullableDateStringSchema,
  riskLevel: riskLevelSchema.or(z.string()),
  monitoringFrequency: monitoringFrequencySchema.or(z.string()),
  lastCheckDate: nullableDateStringSchema,
  nextCheckDate: nullableDateStringSchema,
  dynamicsDescription: z.string().nullable().optional(),
  watchOwnerId: z.number().nullable().optional(),
  watchOwnerLogin: z.string().nullable().optional(),
  attachmentsJson: z.string().nullable().optional(),
  createdAt: z.string(),
  updatedAt: z.string(),
  updatedBy: z.number().int(),
  requiresCheck: z.boolean().optional(),
});

export type WatchlistEntry = z.infer<typeof watchlistEntrySchema>;

// Paginated watchlist response
export const watchlistListResponseSchema = createPaginatedResponseSchema(watchlistEntrySchema);

export type WatchlistListResponse = z.infer<typeof watchlistListResponseSchema>;

// Create watchlist entry request
export const createWatchlistEntrySchema = z.object({
  fullName: z.string().min(1, 'Full name is required'),
  roleStatus: z.string().optional(),
  riskSphereId: z.number().optional(),
  threatSource: z.string().optional(),
  conflictDate: z.string().optional(),
  riskLevel: riskLevelSchema.or(z.string()),
  monitoringFrequency: monitoringFrequencySchema.or(z.string()),
  nextCheckDate: z.string().optional(),
  dynamicsDescription: z.string().optional(),
  watchOwnerId: z.number().optional(),
  attachmentsJson: z.string().optional(),
});

export type CreateWatchlistEntry = z.infer<typeof createWatchlistEntrySchema>;

// Record check request
export const recordCheckRequestSchema = z.object({
  nextCheckDate: z.string().optional(),
  dynamicsUpdate: z.string().optional(),
  newRiskLevel: riskLevelSchema.or(z.string()).optional(),
});

export type RecordCheckRequest = z.infer<typeof recordCheckRequestSchema>;

// Watchlist statistics response
export const watchlistStatisticsSchema = z.object({
  totalEntries: z.number().int().nonnegative(),
  byRiskLevel: z.record(z.string(), z.number()),
  byMonitoringFrequency: z.record(z.string(), z.number()),
  requireingCheck: z.number().int().nonnegative(),
});

export type WatchlistStatistics = z.infer<typeof watchlistStatisticsSchema>;
